Feature/dbt deps tarball

timle2 opened the pull request:
resolves #4205

add new dbt.deps type: url to internally hosted tarball #420

Continued from
https://github.com/dbt-labs/dbt-core/pull/4220

*Revision 3 added Nov 6 2022*

Proposed solution for feature request [4205](https://github.com/dbt-labs/dbt-core/issues/4205)

### Description

Enable direct linking to tarball urls in `packages.yml`, for example:
```
# manufactured test, since you'd want to use hub to install these 
# public tarball used here as example only! 
# this would usually be a tarball hosted  on an internal network
packages:
  - tarball: https://codeload.github.com/dbt-labs/dbt-utils/tar.gz/0.6.5
    name: 'dbt_utils_065'

```
<img width="235" alt="image" src="https://user-images.githubusercontent.com/105436/200200311-f1bd4c51-909c-4a56-8db4-29a63fea75c7.png">
<img width="315" alt="image" src="https://user-images.githubusercontent.com/105436/200200321-8546bc9c-9a2c-4f4a-82e0-86f09855c7b2.png">

Rational:
- dbt projects being self hosted on larger enterprise environments often don't have a connection to the internet (dbt hubs won't work).
- dbt users on larger enterprise environments like to build internal private packages for non-public use (help out other dbt users in company with specific functionality)
-  git package install is not a good option at scale for larger enterprise environments
- internal file hosting service (such as internal artifactory service or internal cloud storage buckets) can be easily configured to host packages for install during deployment, so lets give dbt users a way to install from a direct tar file link


Sketching out doc changes here: 
https://github.com/timle2/docs.getdbt.com/blob/dbt-docs-tarball-package-updates/website/docs/docs/building-a-dbt-project/package-management.md#tar-files


### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose)~, or docs changes are not required/relevant for this PR~
    - https://github.com/dbt-labs/docs.getdbt.com/issues/2474

timle2 commented:
TODO: https://github.com/dbt-labs/dbt-core/pull/4220#discussion_r793784822

> We've updated how to we log to use the Events modules. You'll need to create a new log event for this, and everywhere else you log, and use `fire_event` to kick off writing the log. We have a [`README`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/events/README.md) with some good info on our logging system.

timle2 commented:
TODO
https://github.com/dbt-labs/dbt-core/pull/4220#discussion_r793801205
> As a standard we don't use `assert` in our code other than for tests. They shadow the actual exceptions and throw an `AssertionError` instead. Please rewrite the assert statements to catch relevant exceptions (create new ones if needed!).
> 
> In this specific case, I do like the use `is_tarfile` to check the file! Can you think of a reason it would be worth retrying the download if the file is not a valid tarfile?

emmyoop commented:
@cla-bot check

emmyoop commented:
@timle2 thanks for opening up the PR with your new username!  I've double checked and you did indeed sign the CLA.  We have your username listed in the correct format.  Looks like you may need to configure with your email to get our bot to process it.

> Thank you for your pull request and welcome to our community. We could not parse the GitHub identity of the following contributors: **Tim Leonard**. This is most likely caused by a git client misconfiguration; please make sure to:
> 
> 1. check if your git client is configured with an email to sign commits `git config --list | grep email`
> 2. If not, set it up using `git config --global user.email email@example.com`
> 3. Make sure that the git commit email is configured in your GitHub account settings, see https://github.com/settings/emails

timle2 commented:
> TODO: [#4220 (comment)](https://github.com/dbt-labs/dbt-core/pull/4220#discussion_r793784822)
> 
> > We've updated how to we log to use the Events modules. You'll need to create a new log event for this, and everywhere else you log, and use `fire_event` to kick off writing the log. We have a [`README`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/events/README.md) with some good info on our logging system.

✅ Switched all the logging over to the Events / fire_event. 

> TODO [#4220 (comment)](https://github.com/dbt-labs/dbt-core/pull/4220#discussion_r793801205)
> 
> > As a standard we don't use `assert` in our code other than for tests. They shadow the actual exceptions and throw an `AssertionError` instead. Please rewrite the assert statements to catch relevant exceptions (create new ones if needed!).
> > In this specific case, I do like the use `is_tarfile` to check the file! Can you think of a reason it would be worth retrying the download if the file is not a valid tarfile?

✅ Moved over to properly defined exceptions. For the most part leveraging dbt.exceptions.DependencyException. 

`Can you think of a reason it would be worth retrying the download if the file is not a valid tarfile?` No I think straight failure without retry makes the most sense. In almost all cases the user would want to know that the tarfile is invalid, with `dbt deps` failing, as opposed to hitting the server multiple times to confirm. A transmission error could result in a corrupted tar file, however this is such an edge case I'm not sure it's worth guarding against.

emmyoop commented:
> `Can you think of a reason it would be worth retrying the download if the file is not a valid tarfile?` No I think straight failure without retry makes the most sense. In almost all cases the user would want to know that the tarfile is invalid, with `dbt deps` failing, as opposed to hitting the server multiple times to confirm. A transmission error could result in a corrupted tar file, however this is such an edge case I'm not sure it's worth guarding against.

I agree that this is rare, but it does happen.  The user will still find out the tar file is invalid if that is indeed the issue and not a transmission error we could recover from.  This will be rare but worth protecting against.  In dbt Cloud deps are installed every time a job runs so we want to ensure if there is a transmission error we can recover from with a simple retry, we do so as not to fail an entire job.

Unrelated: You'll want to pull in the latest on the main branch as we've added some auto formatting/checking and that is why you have a failing test.  You can read more about it [here](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#pre-commit).

emmyoop commented on core/dbt/events/types.py:
```diff
@@ -2384,6 +2384,30 @@ def message(self) -> str:
 
+@dataclass
+class Sha1ChecksumPasses(DebugLevel):
+    code: str = "Z049"
```
Love that you added some log lines!  Let's label these as type 'M' instead of 'Z' since it's related to deps generation.

timle2 replied:
thanks for the direction here, made both of these changes

timle2 commented on core/dbt/events/types.py:
```diff
@@ -2384,6 +2384,30 @@ def message(self) -> str:
+class UntarProjectRoot(DebugLevel):
+    subdirectory: str
+    tar_dir_name: str
+    code: str = "Z050"
```
thanks for the direction here, made these changes

timle2 commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,238 @@
+            This simplifies tempfile handling, and prevents multiple hits to
+              url, since we need the file for two seperate operations. One
+              for metadata fetch (in style of other packages) and once for
+              for install for install. '''
```
I've removed caching the tarfile as part of the TarballPinnedPackage, and now the url is hit twice (once to get metadata about the package, and a second time to download the package). It's only a minor changes to code, but if you think that's the right tradeoff, then no problem for me. 

For the more complex topic of reusing registry or git package code - 
I've spent quite a lot of time trying to recycle the registry code, prior to coming to the current solution. However there are some fundamental differences in how this is handled vs registry. Registry has all the lovely json metadata that coordinates the tar download, and naming (as well as enforcing tar file structure, and not allowing subdirectories. Nothing to recycle from registry client, since it's all loading and parsing the json. And the registry install is super simple, since registry api/json handles any unknowns. File source structure and destination name are taken care of, and there is no concept of package subdirectory.
Git provides a good 'cousin' example, in that it allows for subdirectories, and has no metadata source like registry. I've recycled the subdirectory scanning from the git code. However git uses clone and copy for the download and install, so none of the install routine can be copied here. 

Here's a *rough* sketch comparing registry install vs tarball install. Let me know where you have any ideas on streamlining the code between the two. My approach so far has been built on the premise that the registry metadata creates a very simple solution, and so the registry install routine is very straight forward, but the registry install routine overfits to scenario of not needing to 'discover' any metadata. Tarball, on the other hand, given added unknowns from not having metadata, becomes unique in having to discover the metadata from the file structure, and then in turn use that to unpack the package in an intelligent way. Additionally we are doing tarfile verification, sha1 verification, and size verification, which isn't happening for any of the registry (or git, or local) routine, so new code having to be added there. 
![Untitled drawing](https://user-images.githubusercontent.com/105436/155035578-41d5afc8-acbc-4fb7-84cf-16bd64104e38.png)

Phew! ok lots there, but my tldr is please let me know where you see opportunity to streamline/recycle from registry install routine. I've given this a lot of thought already, and do not see any obvious patterns, aside from requiring a third style of install (tarball) that is more of a hybrid of of parts of registry and Git install, than overly similar to either.

timle2 replied:
I've just finished a total rewrite with this `If this behaved the same as the registry package more of the code could be shared to have a single path to follow.` in mind. So ready for a new review :)

emmyoop commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,249 @@
+
+        def handle_local_tar(self):
+            # pass through local tarfile path primarily for testing
+            validate_checksum(self, self.tarball)
+            validate_tarfile(self, self.tarball)
+            tar = tarfile.open(self.tarball, "r:gz")
+            self.tarfile_bin = tar
```
This is duplicated by part of the code in `handle_remote_tar` (#L102-L105).  Combining this will also remove the need for the nested functions as they won't be called more than once.

timle2 replied:
Ok - updated accordingly, see 1dcd97bbc8176060d10c849ed584550ab3f410bd

emmyoop commented on core/dbt/clients/system.py:
```diff
@@ -478,6 +478,24 @@ def untar_package(tar_path: str, dest_dir: str, rename_to: Optional[str] = None)
+def untar_tarfile(
+    TarFile: tarfile.TarFile, dest_dir: str, rename_to: Optional[str] = None
+) -> None:
+    tar_dir_name = dbt.clients.system.resolve_tar_dir_name(TarFile)
```
You can update `untar_package` to use `resolve_tar_dir_name` which will remove the need for the new `untar_tarfile`.   You would just pass through the tar path instead of the TarFile object.

timle2 replied:
Based on your feedback I've removed everything I initially wrote around passing the TarFile object around, in preference for using a tar path. I've removed this `untar_tarfile` function, and moved over to untar_package() as you suggest.

emmyoop commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,249 @@
+
+            fire_event(Sha1ChecksumPasses())
+
+    def download_tar(self, download_url, tar_path):
+        """
+        Sometimes the download of the files fails and we want to retry.  Sometimes the
+        download appears successful but the file did not make it through as expected
+        (generally due to a github incident).  Either way we want to retry downloading
+        and untarring to see if we can get a success.  Call this within
+        `_connection_exception_retry`
+        """
+
+        system.download(download_url, tar_path)
```
This doesn't need to be a function since all it does it this single line.  Directly call `system.download` instead.

timle2 replied:
makes perfect sense, updated accordingly.

the-chris-mitchell commented:
This would be quite useful. Do you still plan to finish this?

timle2 commented:
Yes this is something my team would benefit from as well, and I'm glad others see the value in an additional install method for deps. 
I haven't had time to get this past the finish line unfortunately. 
I welcome any collaborators/input/help whatever it takes to get this approved though! 
At minimum, I'll be able to finish this in Sept, but obviously if others are interested I'd love to get this done sooner.

emmyoop commented:
@timle2 it's great that you're still interested in getting this over the finish line, even if it has to wait until September.  Thanks so much!  I did want to drop in to say that we have done quite a bit of work around dbt deps since you worked on this issue last.  It would be helpful to peruse the changes we've made to (including adding caching!) to ensure your approach matches the pattern we're now using with other methods.

timle2 commented:
@emmyoop 
Based on feedback from revision 1 and 2, I've stripped this PR way down, to the abosolute minumum required to support the tarball feature.
I've removed features - nice but not required - such as supporting subfolders like git packages, and sha1 checks.
Now we lean on existing code from RegistryPinnedPackage (install() and download_and_untar()). This slims down the PR tremendously!
However, RegistryPinnedPackage install() and download_and_untar() are not currently set up as functions that can be used across classes, and I'm not smart enought to set up some crazy class inheritance structure to get it done. So I've copy/pasted the fns from RegistryPinnedPackage to TarballPinnedPackage. I'd advocate for moving these fns to dbt.deps.base, or to a dbt.deps.common file, but I also don't feel that I'm the person to make that call - need some feedback from dbt maintainers here on how to move forward based dbt labs thoughts.

timle2 commented:
Thank you  @the-chris-mitchell for sharing support on this. I've picked up 

> @timle2 it's great that you're still interested in getting this over the finish line, even if it has to wait until September. Thanks so much! I did want to drop in to say that we have done quite a bit of work around dbt deps since you worked on this issue last. It would be helpful to peruse the changes we've made to (including adding caching!) to ensure your approach matches the pattern we're now using with other methods.

done and done! totally rewrote from main branch :)

timle2 commented:
Thank you @the-chris-mitchell  for sharing support on this. I've picked it back up and hoping to get across finish line in the next few weeks.

the-serious-programmer commented:
Would love to see this released! :) (Also see my comment here: https://github.com/dbt-labs/dbt-core/issues/4205#issuecomment-1293401553)

timle2 commented on core/dbt/deps/base.py:
```diff
@@ -97,6 +99,48 @@ def get_installation_path(self, project, renderer):
+        """
+        verbatim from dbt.deps.registry.RegistryPinnedPackage
+            install() from RegistryPinnedPackage class
+            Q for dbt labs, is this an appropriate pattern to refactor against?
```
Questions for dbt labs, is this an appropriate pattern to refactor against?
Fns used upstream in RegistryPinnedPackage and TarballPinnedPackage via inheritance (originally from from dbt.deps.registry.RegistryPinnedPackage)

emmyoop replied:
Sure is!  Everywhere else that inherits from `PinnedPackage` has a custom `install` method.

timle2 replied:
ty for feedback, will close this comment and remove from code

timle2 commented on core/dbt/deps/registry.py:
```diff
@@ -60,32 +57,7 @@ def _fetch_metadata(self, project, renderer) -> RegistryPackageMetadata:
         return RegistryPackageMetadata.from_dict(dct)
 
     def install(self, project, renderer):
-        metadata = self.fetch_metadata(project, renderer)
```
moved upstream to PinnedPackage class so they can be shared across TarballPinnedPackage and RegistryPinnedPackage via PinnedPackage inheritance.

timle2 commented:
@emmyoop this is ready for another run against the git workflows for tests/approvals, can you please trigger? Previously failed due to formatting errors, which are fixed now (bf43d1f).

timle2 commented:
@emmyoop something funny going on with GitHub actions today? Last git workflow run failed with the following:
<img width="1083" alt="image" src="https://user-images.githubusercontent.com/105436/203144188-1ee3d981-34e6-4d75-8dfb-9b13c4d22f4e.png">

dbeatty10 commented:
@timle2 I think branches that start with `feature/` have a special meaning for us.

Can you rename it to something else (like `timle2/dbt-deps-tarball`) and see if it resolves the issue you are seeing with `pre-commit`?

timle2 commented:
> @timle2 I think branches that start with `feature/` have a special meaning for us.
> 
> Can you rename it to something else (like `timle2/dbt-deps-tarball`) and see if it resolves the issue you are seeing with `pre-commit`?

Thanks for the help here. I've got a second PR with the altered branch name here:
https://github.com/dbt-labs/dbt-core/pull/6302

timle2 commented:
> > @timle2 I think branches that start with `feature/` have a special meaning for us.
> > Can you rename it to something else (like `timle2/dbt-deps-tarball`) and see if it resolves the issue you are seeing with `pre-commit`?
> 
> Thanks for the help here. I've got a second PR with the altered branch name here, see if that works. #6302 No way for me to change the branch name here without closing the PR it seems.

Actually this PR working fine now, yay! Back to getting flake errors like I'm used to 😆

dbeatty10 commented:
@timle2 Something for you to do:
1. Add this as the first line of your PR description:
    - `resolves #4205`
    - The ["resolves" keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) will link the PR to the issue and will automatically close the latter at the appropriate time.
    - Having it as the first line is just a convention we used and not strictly required (but appreciated!)

emmyoop commented on core/dbt/task/deps.py:
```diff
@@ -37,10 +37,13 @@ def track_package_install(
     ) -> None:
         # Hub packages do not need to be hashed, as they are public
-        # Use the string 'local' for local package versions
+        # Use the string 'local' for local and 'tarball' package versions
```
This comment doesn't really add value, you can delete the entire line instead of modifying it.

timle2 replied:
removed with 3d48f6ad612299db7d3d92464d2df1a518f991f7

emmyoop reviewed (commented):
@timle2 this looks great! 🎉   Just had a few small comments.  

I wanted to give you some feedback now since I know it was overdue!  Before I officially approve and merge this in I wanted to do some manual testing on it.  I'll let you know if I find anything testing and if not I'll get this merged in once you finish a bit of cleanup.

timle2 commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,249 @@
+                   f"of {TARFILE_MAX_SIZE}.")
+            raise DependencyException(msg)
+
+            fire_event(Sha1ChecksumPasses())
```
This (and the sha1 checks) were removed as of [25ee2c6](https://github.com/dbt-labs/dbt-core/pull/4689/commits/25ee2c69712186dc3184fa966b66a318bd82e37b). There's no matches for `fire_event(Sha1ChecksumPasses())` in the current revision. Looks like this may just be based on an outdated snippet, but let me know if I've missed something!

timle2 commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,90 @@
+        return self.version
+
+    def nice_version_name(self):
+        return "version {}".format(self.version)
```
ty! will keep this in mind going forward.

timle2 commented:
> @timle2 this looks great! 🎉 Just had a few small comments.
> 
> I wanted to give you some feedback now since I know it was overdue! Before I officially approve and merge this in I wanted to do some manual testing on it. I'll let you know if I find anything testing and if not I'll get this merged in once you finish a bit of cleanup.

🥳 Thanks @emmyoop! Thanks for the support through this PR, greatly appreciated. I've addressed all the comments, but please let me know if I've missed anything. 

Flake/Black passing on my local machine, so this should be good to re-run workflows for another round of validation.

dbeatty10 commented:
I'll do some manual testing on this and report back here afterwards.

Mathyoub commented:
@dbeatty10 and I did some testing and the happy path works great! But, we would like to request some improvements or at least raise some observations for discussion. 

Because version is optional and can lead to misleading output we would recommend removing it altogether. In the case of a user specifying a different version than the actual tarball, we will see a package.yml that looks like this: 
```
packages:
  - tarball: https://codeload.github.com/dbt-labs/dbt-utils/tar.gz/0.6.5
    name: 'dbt_utils_065'
    version: 0.9.6
```
Produce an output like this:
```
23:43:57  Installing dbt-labs/dbt_utils
23:43:58    Installed from version 0.9.6
23:43:58    Up to date!
```
Where `Installed from version 0.9.6` is actually incorrect. Another clear example would be passing version but not specifying version and getting `Installed from version None`.

Another possibly improvement can be seen when a user give a name, no version, but tries something else like `revision:` or maybe a typo `versoin:` the output is not exactly user friendly and returns a stack trace that doesn't help the user much. Example: 
```
packages:
  - tarball: https://codeload.github.com/dbt-labs/dbt-utils/tar.gz/0.6.5
    name: 'dbt_utils_065'
    versoin: 0.9.6
```
Gives us an output: 
```
00:04:17  Encountered an error:
'version'
00:04:17  Traceback (most recent call last):
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/main.py", line 135, in main
    results, succeeded = handle_and_check(args)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/main.py", line 198, in handle_and_check
    task, res = run_from_args(parsed)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/main.py", line 225, in run_from_args
    task = parsed.cls.from_args(args=parsed)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/task/deps.py", line 99, in from_args
    return super().from_args(args)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/task/base.py", line 109, in from_args
    config = cls.ConfigType.from_args(args)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/runtime.py", line 613, in from_args
    project, profile = cls.collect_parts(args)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/runtime.py", line 208, in collect_parts
    project = cls.collect_project(args=args, project_renderer=project_renderer)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/runtime.py", line 246, in collect_project
    project = partial.render(project_renderer)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/project.py", line 286, in render
    return self.create_project(rendered)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/project.py", line 408, in create_project
    packages = package_config_from_data(rendered.packages_dict)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/config/project.py", line 106, in package_config_from_data
    PackageConfig.validate(packages_data)
  File "/Users/dbeatty/projects/dbt-core-4689/core/dbt/contracts/project.py", line 103, in validate
    if not package["version"]:
KeyError: 'version'
```
This case could be handled similarly to how this resolves:
```
packages:
	- git: "https://github.com/dbt-labs/dbt-utils.git"
	  version: 1.0.0
```
```
16:40:47  Encountered an error while reading the project:
16:40:47  ERROR: Runtime Error
	The packages.yml file in this project is malformed. Please double check
	the contents of this file and fix any errors before retrying.

	You can find more information on the syntax for this file here:
	https://docs.getdbt.com/docs/package-management

	Validator Error:
	{'git': 'https://github.com/dbt-labs/dbt-utils.git', 'version': '1.0.0'} is not valid under any of the given schemas

Error encountered in /Users/matthewbeall/dbt-projects/propmgmt/dbt_project.yml
16:40:47  Encountered an error:
Runtime Error
	Could not run dbt
```
In this example revision is expected instead of version and gives a nice message about where and what the error is as well as reference for syntax.

We should also ensure [this doc](https://docs.getdbt.com/docs/build/packages#how-do-i-add-a-package-to-my-project)  is updated eventually.

emmyoop commented:
@Mathyoub thanks for the great testing and feedback!  I agree that `version` isn't necessary.  At a quick glance I don't see why the TarballPackage isn't hooked in to validation, but we definitely want to get the helpful validation error on invalid package definitions.

@timle2 I just want to add that you only need to open an issue for a new feature [here](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) and a PM will get the docs written up.

However, adding info on `tarball.py` to the [README](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/deps/README.md) would be perfect.

dbeatty10 commented:
> @timle2 I just want to add that you only need to open an issue for a new feature [here](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) and a PM will get the docs written up.

I just created that issue [here](https://github.com/dbt-labs/docs.getdbt.com/issues/2474) so one less thing that needs to be done.

Updated the PR description accordingly.

timle2 commented:
@Mathyoub Really appreciate the feedback here, and very much appreciate forcing me off the `happy path` 😄 

I'm submitting fixes to your two points with 7cb8a4ab5e3bb5c33bdfa72e28ea7214a8e3fd99. More context on each below: 

1) **No need for version** Totally agree. The registry install (now moved to `dbt.deps.base PinnedPackage _install`) expects a version entry, so that's why I included it originally.  
To address your comments I've opted to set the version to in the tarball class as "". I'm reluctant to remove `version` and `get_version` from the tarball class, since  there's a bunch of places dbt expects the pinned packaged class to return a usable .version or .version_name, so I'm inclined to find a solution that leaves it in the class (e.g. version = ""). For example `dbt.deps.base PinnedPackage _install`, or below
<img width="471" alt="image" src="https://user-images.githubusercontent.com/105436/205464881-6439e0d9-2f6e-444c-b47d-dc9716e6ca91.png">


2) **Bad stack trace no good 100%** Great news, the changes from (1) fixed it!  
There is something brittle about using optional entries in core.dbt.contracts.project (and my shallow understanding of the full dbt code base).  
Somehow, the way I set it up, caused that `line 103, in validate \n\t if not package["version"]:` error, instead of the nicer `MALFORMED_PACKAGE_ERROR` you referenced.  
It's a moot point though: removing `version` (as in 1) -> removing the optional entry from `core.dbt.contracts.project TarballPackage` -> `MALFORMED_PACKAGE_ERROR` being properly called. So yeah, fixed (in the context of this PR at least)! 

**Now working examples of the cases you raised:**
*happy path*
Note: using the packages.yml tarball url now in tarball.get_version so that this message returns something informative to the user.
<img width="749" alt="image" src="https://user-images.githubusercontent.com/105436/205465058-1c9cd56e-bf1f-4b2e-bbea-cf1c67fa788e.png">

*duplicate project names in packages.yml*
<img width="775" alt="image" src="https://user-images.githubusercontent.com/105436/205464821-3b759d0f-51f2-4fd9-9fcd-ba551c9ddf33.png">

*version removed*
Note: helpful error appears, is as expected.
<img width="760" alt="image" src="https://user-images.githubusercontent.com/105436/205465120-692e4b4a-3c8c-4df5-b752-b52ff6084192.png">

*unexpected field added*
Note: helpful error appears, is as expected.
<img width="755" alt="image" src="https://user-images.githubusercontent.com/105436/205465176-bc8ce47c-557d-4ba5-9901-b8987ee65dbf.png">

timle2 commented:
>

timle2 commented:
> @timle2 I just want to add that you only need to open an issue for a new feature [here](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) and a PM will get the docs written up.
> 
> However, adding info on `tarball.py` to the [README](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/deps/README.md) would be perfect.

@emmyoop Thanks for flagging this! I've added/altered the readme with dc8ba0a30cc042df8d5dc9e166990cfba607ab75

iknox-fa commented on core/dbt/deps/base.py:
```diff
@@ -97,6 +99,34 @@ def get_installation_path(self, project, renderer):
+    def _install(self, project, renderer):
+        metadata = self.fetch_metadata(project, renderer)
+
+        tar_name = "{}.{}.tar.gz".format(self.package, self.version)
```
Minor detail:  We prefer using the more modern [f-string style python string formatting](https://peps.python.org/pep-0498/).

timle2 replied:
Thanks Ian - I've changed it over to f-string with 8787ba41af585b954d18c24c2dccbf147035cd9f

iknox-fa commented on core/dbt/deps/base.py:
```diff
@@ -97,6 +99,34 @@ def get_installation_path(self, project, renderer):
+        metadata = self.fetch_metadata(project, renderer)
+
+        tar_name = "{}.{}.tar.gz".format(self.package, self.version)
+        tar_path = os.path.realpath(os.path.join(get_downloads_path(), tar_name))
```
Minor detail:  We're trying to avoid using `os.path` for new commits to dbt-core (preferring [Pathlib](https://docs.python.org/3/library/pathlib.html) instead).  THis is due to the many odd caveats when running in non-posix environments.

iknox-fa reviewed (approved):
LGTM other than the few mentioned minor details.

Mathyoub commented on core/dbt/deps/tarball.py:
```diff
@@ -0,0 +1,74 @@
+        return self.version
+
+    def nice_version_name(self):
+        return f"Tarball (url: {self.tarball})"
```
Can we have this be "tarball" instead of "Tarball"?

timle2 replied:
changed, thanks! 5410f1f6188bd87efebadc86c7d8cd9545aca744

Mathyoub commented:
Changes look great @timle2! Just left one minor comment to lowercase "tarball" in the output!

timle2 commented:
Tests failing with https://github.com/dbt-labs/dbt-core/pull/4689/commits/b06a66240a5d391b66577d298fc08fa43704f1d8 due to tests in test/unit/test_deps.py that still referenced a version input to tarball class. Fixed with https://github.com/dbt-labs/dbt-core/pull/4689/commits/e9f2121f9dc5cc2adc38082f635295beed321401. pytest passing now on local machine.

dbeatty10 commented:
@timle2 we'll do some investigation and troubleshooting on our end for the two CI tests that aren't passing:
- Generate CLI API docs / check if generation needed
- Tests and Code Checks / integration test / python 3.8 / windows-latest

<img width="600" alt="image" src="https://user-images.githubusercontent.com/44704949/205790392-b9e44d5f-4120-4eda-bac9-907c82d87cc1.png">
<img width="600" alt="image" src="https://user-images.githubusercontent.com/44704949/205790478-fa55fdf8-502e-48d5-961e-d577f38a3550.png">

dbeatty10 commented:
> @timle2 we'll do some investigation and troubleshooting on our end for the two CI tests that aren't passing:
> 
> * Generate CLI API docs / check if generation needed
> * Tests and Code Checks / integration test / python 3.8 / windows-latest

👍 After investigation, we are good from the perspective of CI tests.

### Justification
Confirmed with @emmyoop and @stu-k that `Generate CLI API docs` can be ignored for this PR. Alternatively, if you do fresh merge with `main`, it will skip this particular check in the future.

The windows-latest `Tests and Code Checks` was flaky and addressed by just manually re-running it. A future issue/PR combo with [these](https://github.com/dbt-labs/dbt-core/pull/6346/files#diff-5bab13c218ba16b33ff49936fb3b417cce1da20f705a4010d3e07d566100428c) updates should be able to address that flakiness.

timle2 commented:
@emmyoop I think we are complete on the PR. Lmk if there is anything needed from me before Merge!

emmyoop reviewed (approved):
@timle2 thank you again for all your work (and persistence) on this!  This is such a great community contribution!  We'll be including it in the 1.4 release!

timle2 commented:
> @timle2 thank you again for all your work (and persistence) on this! This is such a great community contribution! We'll be including it in the 1.4 release!

Amazing! Thanks to you as well for your patience and guiding me through all the first time contributor stuff. Looking forward to making some more (and much smoother) contributions in the future!

jtcohen6 commented:
Nice work team !!
